File: src/services/ats/workable.ts
import { logger } from "../../logger";
import {
  type AtsJobData,
  type Fetcher,
  type WorkableJob,
  type WorkplaceType,
  workableListResponseSchema,
} from "./types";

const log = logger.child({ component: "ats/workable" });

const LIST_ENDPOINT = "https://apply.workable.com/api/v3/accounts";

export function parseWorkableUrl(url: string): { slug: string; shortcode: string } | null {
  try {
    const { hostname, pathname } = new URL(url);
    if (!hostname.includes("workable.com")) return null;
    const segments = pathname.split("/").filter(Boolean);
    // apply.workable.com/{slug}/j/{shortcode}/?
    const jIdx = segments.indexOf("j");
    if (jIdx <= 0 || jIdx === segments.length - 1) return null;
    const slug = segments[jIdx - 1];
    const shortcode = segments[jIdx + 1];
    if (!slug || !shortcode) return null;
    return { slug, shortcode };
  } catch {
    return null;
  }
}

function normalizeWorkplaceType(value: string | null | undefined): WorkplaceType | null {
  if (!value) return null;
  switch (value.toLowerCase()) {
    case "remote":
      return "Remote";
    case "hybrid":
      return "Hybrid";
    case "on_site":
    case "onsite":
    case "on-site":
      return "OnSite";
    default:
      return null;
  }
}

function formatLocation(loc: WorkableJob["location"]): string {
  if (!loc) return "";
  const parts = [loc.city, loc.country].filter((s): s is string => !!s);
  return parts.join(", ");
}

/**
 * Workable's list endpoint paginates 10 results at a time and we have no
 * single-job endpoint, so we use full-text `query` on the title to land the
 * target on the first page. The shortcode (extracted from the URL) is the
 * ground-truth identifier — we filter the results by it. If the query misses
 * (title got mangled, or the job is on page 2+), we return null and the
 * pipeline falls back to body-only evaluation, same as it does today.
 */
export async function fetchWorkableJob(
  url: string,
  query: string,
  fetcher: Fetcher = fetch,
): Promise<AtsJobData | null> {
  const parsed = parseWorkableUrl(url);
  if (!parsed) return null;
  const { slug, shortcode } = parsed;

  const apiUrl = `${LIST_ENDPOINT}/${slug}/jobs`;

  let res: Response;
  try {
    res = await fetcher(apiUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json", Accept: "application/json" },
      body: JSON.stringify({ query }),
    });
  } catch (err) {
    log.warn({ err, url }, "workable fetch failed");
    return null;
  }

  if (!res.ok) {
    log.warn({ status: res.status, url }, "workable non-ok response");
    return null;
  }

  let raw: unknown;
  try {
    raw = await res.json();
  } catch (err) {
    log.warn({ err, url }, "workable json parse failed");
    return null;
  }

  const result = workableListResponseSchema.safeParse(raw);
  if (!result.success) {
    log.warn({ url, issues: result.error.issues }, "workable schema mismatch");
    return null;
  }

  const job = result.data.results.find((j) => j.shortcode === shortcode);
  if (!job) {
    log.warn(
      { slug, shortcode, query, returned: result.data.results.length },
      "workable shortcode not found in query results",
    );
    return null;
  }

  const primary = formatLocation(job.location);
  const secondary = (job.locations ?? []).map(formatLocation).filter((s) => s.length > 0);
  const locations = primary && !secondary.includes(primary) ? [primary, ...secondary] : secondary;

  return {
    source: "workable",
    location: primary,
    locations,
    workplaceType: normalizeWorkplaceType(job.workplace),
    country: job.location?.country ?? null,
  };
}

File: job_finder/ats/workable.py
from __future__ import annotations

from typing import ClassVar

from pydantic import BaseModel, ConfigDict

from job_finder.ats.models import AtsAvailable, normalize_workplace_type, unique_locations
from job_finder.urls import parse_http_url


class _WorkableWireModel(BaseModel):
    model_config: ClassVar[ConfigDict] = ConfigDict(strict=True)


class _WorkableLocation(_WorkableWireModel):
    country: str | None = None
    city: str | None = None


class _WorkableJob(_WorkableWireModel):
    shortcode: str
    workplace: str | None = None
    location: _WorkableLocation | None = None
    locations: list[_WorkableLocation] | None = None


class _WorkableResponse(_WorkableWireModel):
    results: list[_WorkableJob]


def parse_workable_url(url: str) -> tuple[str, str] | None:
    parsed = parse_http_url(url)
    if parsed is None or parsed.hostname is None:
        return None
    if parsed.hostname != "apply.workable.com":
        return None
    segments = tuple(segment for segment in parsed.path.split("/") if segment)
    try:
        job_index = segments.index("j")
    except ValueError:
        return None
    if job_index == 0 or job_index == len(segments) - 1:
        return None
    return segments[job_index - 1], segments[job_index + 1]


def parse_workable_job(payload: object, shortcode: str) -> AtsAvailable | None:
    response = _WorkableResponse.model_validate(payload)
    job = next(
        (candidate for candidate in response.results if candidate.shortcode == shortcode), None
    )
    if job is None:
        return None
    primary = _format_location(job.location)
    secondary = tuple(
        location for item in job.locations or () if (location := _format_location(item))
    )
    return AtsAvailable(
        source="workable",
        location=primary,
        locations=unique_locations(primary, secondary),
        workplace_type=normalize_workplace_type(
            job.workplace,
            onsite_values=("onsite", "on-site", "on_site"),
        ),
        country=job.location.country if job.location else None,
    )


def _format_location(location: _WorkableLocation | None) -> str:
    if location is None:
        return ""
    return ", ".join(part for part in (location.city, location.country) if part)
